Action scripts - AngerNodes manipulation

FS-21 opened the pull request:
This pack of actions manipulates the hate of the houses towards other houses. This don't replace the game function "make_enemies()" but helps the AI to pick the one with highest value for being the most hated (the main enemy of the house used in the AI trigger tag called "OnlyTargetHouseEnemy").

The hate table values of all Houses increase quickly depending of how much time another house is attacking by another house & decreases very slow in moments of peace.

Curiosity note: this can be used for overriding the known MP bug of allied AI with the human player at the beginning of the games that they start as pacifists until they get attacked by others (because their only hated player is the 1º human player and that one is "ally" so they can't attack him).
The same reason explains why initially in the MP battle all AI houses attacks the player.

-----------------------------------

No official documentation yet but I'll  copy here all the details & possible values of each action:


```
Action 114 - Set House anger value modifier
Affects how much hate applies to a selected house (depends of the script action). This anger value is applied only inside the scriptType of the Team.
Format: 114,n
Possible values: Positive for increase hate and negative values for decreasing hate. Default to 0
Affects actions: 116, 117, 118, 119, 120 & 121


Action 115 - Override OnlyTargetHouseEnemy value
The value of the tag OnlyTargetHouseEnemy in AI Triggers can be modified for the new Attack & Move actions. Only affects the next new attack or move action script. These anger values are applied only in the House owner of the team.
Format: 115,n
Possible values: 
-1 ; Use default value specified in OnlyTargetHouseEnemy tag
0  ; force the tag value to FALSE
1  ; force the tag value to TRUE
2  ; force the tag value to TRUE or FALSE randomly.


Action 116 - Modify House hate with index
Modifies a specific House hate using its index value as House array selector. Is applied to the House owner of the team.
Format: 116,n
Possible values:
n >= 0 (keep in mind that a Multiplayer map only has 8 players plus "Neutral & Special" houses) so doesn't have any sense to pick a higer value than the index of the last House. Pobably isn't recommended for all modders if you don't know what index have each player. These anger values are applied only in the House owner of the team.

Action 117 - Modify Hate values from a list of Houses
Picks a list of Houses from the rulesmd.ini section called [AIHousesList] and if any House appear in the scenario then the hate value against those Houses will be modified. These anger values are applied only in the House owner of the team.
Format of the new section:
[AIHousesList]
0=Here_Your_House_ID_1,Here_Your_House_ID_2, ...
1=...
Format: 117,n
Possible values:
>= 0 integer values that is the index of the list in [AIHousesList]


Action 118 - Modify randomly the Hate value of a house specified in a list of Houses
Like action 117 but only picks 1 house randomly from the specified list and then search that selected House in the scenario for midifying the hate value. These anger values are applied only in the House owner of the team.
Format of the new section:
[AIHousesList]
0=Here_Your_House_ID_1,Here_Your_House_ID_2, ...
1=...
Format: 118,n
Possible values:
>= 0 integer values that is the index of the list in [AIHousesList]


Action 119 - Set the most hated House ("<" comparison)
Basicly it increases the owner's anger against an Enemy House converting the selected House as the main enemy of the Team's owner. Depends on the specified argument.
Format: 119,n
Possible values:
-7 ; The House with less house kills is selected
-6 ; The House with less free power (free = production - consumption) is selected
-5 ; The House with less power production is selected
-4 ; The House with less power consumption is selected
-3 ; The nearest enemy Human base is selected
-2 ; The poorest House is selected
-1 ; The enemy House with nearest unit to the Team Leader is selected
>0 ; (using the parameters from the new attack/move script actions) The House with less threat of the checked type (sum of all the units of the same checked type * threat value)


Action 120 - Set the most hated House (">" comparison)
Basicly it increases the owner's anger against an Enemy House converting the selected House as the main enemy of the Team's owner. Depends on the specified argument.
Format: 120,n
Possible values:
-7 ; The House with more house kills is selected
-6 ; The House with more free power (free = production - consumption) is selected
-5 ; The House with more power production is selected
-4 ; The House with more power consumption is selected
-3 ; The farthest enemy Human base is selected
-2 ; The richest House is selected
-1 ; The enemy House with farthest unit to the Team Leader is selected
>0 ; (using the parameters from the new attack/move script actions) The House with more threat of the checked type (sum of all the units of the same checked type * threat value)


Action 121 - Set the most hated House randomly
Basicly it increases the owner's anger against an Enemy House picked randomly.
Format: 121,0


Action 122 - Reset anger against other Houses
Sets to 0 every House hate level. Is applied to the House owner of the team.
Format: 122,0


Action 123 - 
The selected house will become the most hated House of the map (the effects are only visible if the other houses are enemy of the selected house)
Format: 123,n
Possible values:
-3   ; All Human players will be hated by everyone.
-2   ; The Team's owner will be the most hated by everyone.
-1   ; Selects a random House. The own House is excluded in the selection of the most hated by everyone.
>= 0 ; Index of the House that will be hated by everyone.

```

FS-21 committed:
Initial commit

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/script-action-houses-anger

FS-21 committed:
move action 113 to 120

FS-21 committed:
added aggro action

FS-21 committed:
improvements

in SetTheMostHatedHouse() & GetTheMostHatedHouse()

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
new features when finding the most hateful house

FS-21 committed:
debug fixes

FS-21 committed:
Some fixes & improvements

YRIDZE reviewed (dismissed):
Please, review your code for properly formatted "if" long statement separation (most of them I highlighted) and "switch" statements (availability of the required blank lines). And also break up long log lines.

FS-21 committed:
Update src/Ext/Script/Body.cpp

Co-authored-by: YRIDZE <78161538+YRIDZE@users.noreply.github.com>

FS-21 committed:
Update src/Ext/Script/Body.cpp

Co-authored-by: YRIDZE <78161538+YRIDZE@users.noreply.github.com>

FS-21 committed:
Update src/Ext/Script/Body.cpp

Co-authored-by: YRIDZE <78161538+YRIDZE@users.noreply.github.com>

FS-21 committed:
Update src/Ext/Script/Body.cpp

Co-authored-by: YRIDZE <78161538+YRIDZE@users.noreply.github.com>

FS-21 committed:
Update src/Ext/Script/Body.cpp

Co-authored-by: YRIDZE <78161538+YRIDZE@users.noreply.github.com>

FS-21 committed:
Update src/Ext/Script/Body.cpp

Co-authored-by: YRIDZE <78161538+YRIDZE@users.noreply.github.com>

FS-21 committed:
fixes

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Added debug lines

When using ObjectInfo or the AngerNode Action is executed successfuly some debug lines will be written in debug.log with the current AngerNode values of each house.

FS-21 commented:
For the testing of the PR I added some new debug lines that will appear in debug.log each time the action is executed successfuly or ObjectInfo is used.

Data will look like this:

> Player 0 [Observer] 
> Player 1 [GDICountry2]: 0:0, 2:1, 3:0, 4:0 -> Main Enemy House: 2
> Player 2 [AlliesCountry]: 0:1, 1:1, 3:0, 4:0 -> Main Enemy House: 1
> Player 3 [Neutral]: 0:0, 1:0, 2:0, 4:0 -> Main Enemy House: -1
> Player 4 [Special]: 0:0, 1:0, 2:0, 3:0 -> Main Enemy House: -1

How interpret each House's line?:
<HouseClass internal index in the game> [<HouseClass name>]: <enemy index #1:Anger value>, <enemy index #2:Anger value>, ...  -> <Index of the House's enemy>

FS-21 committed:
Remove some useless comments

FS-21 committed:
improvement for unknown actions

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
small fix

FS-21 committed:
Point to latest YRpp

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Update YYpr

FS-21 committed:
removing unnecessary lines

FS-21 committed:
Merge remote-tracking branch 'FS-21/develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Point to latest YRpp

FS-21 committed:
Added AngerNodes documentation

Documentation for Script Actions 114 to 123.

FS-21 committed:
Small information change in action 114

FS-21 commented:
Added the missing documentation & Readme.

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/script-action-houses-anger

FS-21 committed:
Point to latest YRpp

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
point to latest YRpp

FS-21 committed:
Added 3 new comparisons for selecting the most hated house:

-8 ===> The enemy House with more/less naval units is selected.
-9 ===> The enemy House with more/less aircraft docks is selected.
-10 ===> The enemy House with more/less factories is selected (excluded the aircraft factories).

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
point to latest YRpp

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

SMxReaver commented:
Tested out this feature as extensively as I could based off nightly build 02e7afa:

- Script 114 to set hate seems to be working accordingly as written. I have tested it with both positive and negative values, and seen the hate modifiers seem to grow or weaken over time in-game. This was confirmed with targeting a specific house and then the chosen house be the only one present on the map, such as more hate towards Libya or less hate against Korea.

- Script 115 I was unsure of how to fully test, consider it to be **untested** as of this report from my end.

- Script 116 seems to be working if I understand how it works. This allows for targeting specific players with attacks based on their starting point.

- Script 117 also seems to be working as described. I created two unique AIHousesList entries, and combined this script with the tests mentioned in Script 114. 

- Script 118 seems to work. I set two houses on the map that were in an AIHousesList entry, and set two additional houses to be control houses. The randomizer would apply 2500 hate to the randomly chosen house, and then the AIs would focus only on one of those two houses. This will apply the hate modifier to all houses of the chosen script, even if the house is no longer a player (noted with using CnCNet client's Observer player option setting my house to Russia and the script target only increasing "Russians" hate level as coded leading to Observers having hate but not main target)

- Script 119 adapts to changes in player conditions that are tracked by the script pretty well. Can't say I see any large issues here, I did not test every single negative and positive possible condition however.

- Script 120 adapts to changes in player conditions that are tracked by the script pretty well. Can't say I see any large issues here, I did not test every single negative and positive possible condition however. (copypasta from script 119 as they're similar in function)

- Script 121 appears to work, the enemy seems to pick new enemy houses every time out of 10 games observed. 121 does not output debug information into debug.log, I used another new action's reporting on hate levels from this PR to double-check the changes in debug.log

- Script 122 works as intended, the house hate modifiers are constantly reset throughout the game. This is another feature that does not output information to the debug.log file, so I tested by giving a house not present ingame a hate modifier to produce debug.log output to confirm tests were successful.

- Script 123 did not appear to react to positive values or negative values as the condition, perhaps some more information about the use-case would help in testing to make sure that there's functionality or not. Currently, I'd say that **Script 123 is not working.**

With about 90% of this feature working (random rough number) as described, I'd say this feature looks good. I'd like someone else to help follow up with Script 115 and 123 to double confirm they work, but help being what it is!

As a small side note, I had a situation where I was a 'Russia' observer via CnCNet client and 'Script 118 - Modify Hate Randomly' script picked "Russians" as the hated house. Hate levels for the observer player were increased too as declared in debug.log, I do not know if this is an intended behavior or could have unintended consequences in performance for picking targets.

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Point to latest YRpp

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/script-action-houses-anger

FS-21 committed:
updated documentation and other missing bits

Renamed actions:

OverrideOnlyTargetHouseEnemy = 115 -> 14005,
SetHouseAngerModifier = 114 -> 14006,
ModifyHateHouseIndex = 116 -> 14007,
ModifyHateHousesList = 117 -> 14008,
ModifyHateHousesList1Random = 118 -> 14009,
SetTheMostHatedHouseMinorNoRandom = 119 -> 14010,
SetTheMostHatedHouseMajorNoRandom = 120 -> 14011,
SetTheMostHatedHouseRandom = 121 -> 14012,
ResetAngerAgainstHouses = 122 -> 14013,
AggroHouse = 123 -> 14014,

FS-21 committed:
Extra debug info

Will be removed before the final merge.

FS-21 committed:
small fix

due to YRpp changes

FS-21 committed:
.

FS-21 committed:
Fixed bug

Insta-win in Campaign maps.

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Compilation fixes

FS-21 committed:
Tweaks

FS-21 committed:
tweaks and small changes

- Observers should be ignored.
- small tweaks.

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
YRpp

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
YRpp fix

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/script-action-houses-anger

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Moved code into a new file and updated code

FS-21 committed:
Tweak

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+
+void ScriptExt::ResetAngerAgainstHouses(TeamClass* pTeam)
+{
+	// Invalid team
+	if (!pTeam)
+		return;
```
I don't think this will ever be `nullptr` unless something is seriously wrong, and in that case this code not working is least of your worries so I don't think this check is necessary.

FS-21 replied:
If is safe then I'll remove similar lines.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+
+void ScriptExt::SetHouseAngerModifier(TeamClass* pTeam, int modifier = 0)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+	if (!pTeam)
+		return;
+
+	auto pTeamData = TeamExt::ExtMap.Find(pTeam);
+	if (!pTeamData)
```
Add empty line before the if statement.

FS-21 replied:
Already mentioned in another PR about this case is valid in the code style.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+
+void ScriptExt::ModifyHateHouses_List(TeamClass* pTeam, int idxHousesList = -1)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+	if (!pTeam)
+		return;
+
+	auto pTeamData = TeamExt::ExtMap.Find(pTeam);
+	if (!pTeamData)
```
Ditto with if statement and spacing.

FS-21 replied:
Already mentioned. Resolve conversation.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+
+void ScriptExt::ModifyHateHouses_List1Random(TeamClass* pTeam, int idxHousesList = -1)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+	if (!pTeam)
+		return;
+
+	auto pTeamData = TeamExt::ExtMap.Find(pTeam);
+	if (!pTeamData || pTeamData->AngerNodeModifier == 0)
```
Ditto with if statement and spacing.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+
+void ScriptExt::SetTheMostHatedHouse(TeamClass* pTeam, int mask = 0, int mode = 1, bool random = false)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+	if (!pTeam)
+		return;
+
+	auto pTeamData = TeamExt::ExtMap.Find(pTeam);
+	if (!pTeamData)
```
Ditto with if statement and spacing.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+
+HouseClass* ScriptExt::GetTheMostHatedHouse(TeamClass* pTeam, int mask = 0, int mode = 1)
+{
+	if (!pTeam)
+		return nullptr;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+	if (!pTeam)
+		return nullptr;
+
+	auto pTeamData = TeamExt::ExtMap.Find(pTeam);
+	if (!pTeamData || mask == 0)
```
Ditto with if statement and spacing.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+	int enemyStructures = -1;
+	int enemyNavalUnits = -1;
+
+	if (mask == -2) // Based on House economy
```
These if statement blocks contain a lot of duplicated code with the house loops etc. If you could find a way to condense this while not polluting the code with myriad of if statements that would be great.

FS-21 replied:
ScriptExt::GetTheMostHatedHouse(...) was refactored due to the multiple duplicated portions of code you mentioned and I applied other feedback changes.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+// Note: only works for new Phobos script actions, not the original ones
+void ScriptExt::OverrideOnlyTargetHouseEnemy(TeamClass* pTeam, int mode = -1)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+	if (!pTeam)
+		return;
+
+	auto pTeamData = TeamExt::ExtMap.Find(pTeam);
+	if (!pTeamData)
```
Ditto with if statement and spacing.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+
+void ScriptExt::ModifyHateHouse_Index(TeamClass* pTeam, int idxHouse = -1)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+	if (!pTeam)
+		return;
+
+	auto pTeamData = TeamExt::ExtMap.Find(pTeam);
+	if (!pTeamData || pTeamData->AngerNodeModifier == 0)
```
Ditto with if statement and spacing.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+// The selected house will become the most hated of the map (the effects are only visible if the other houses are enemy of the selected house)
+void ScriptExt::AggroHouse(TeamClass* pTeam, int index = -1)
+{
+	if (!pTeam)
+		return;
```
Ditto with the `nullptr` check here.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+	if (!pTeam)
+		return;
+
+	auto pTeamData = TeamExt::ExtMap.Find(pTeam);
+	if (!pTeamData)
```
Ditto with if statement and spacing.

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+		ScriptExt::UpdateEnemyHouseIndex(pHouse);
+	}
+
+	ScriptExt::DebugAngerNodesData(); // Remove this line before merging into develop branch!
```
About time to do it, right?

Starkku commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+	pHouse->EnemyHouseIndex = index;
+}
+
+void ScriptExt::DebugAngerNodesData()
+{
+	ScriptExt::Log("Updated AngerNodes lists of every playable House:\n");
+
+	for (auto pHouse : *HouseClass::Array)
+	{
+		if (pHouse->IsObserver())
+			ScriptExt::Log("Player %d [Observer] ", pHouse->ArrayIndex);
+		else
+			ScriptExt::Log("Player %d [%s]: ", pHouse->ArrayIndex, pHouse->Type->ID);
+
+		int i = 0;
+
+		for (auto& angerNode : pHouse->AngerNodes)
+		{
+			if (!pHouse->IsObserver())
+				ScriptExt::Log("%d:%d", angerNode.House->ArrayIndex, angerNode.AngerLevel);
+
+			if (i < HouseClass::Array->Count - 2 && !pHouse->IsObserver())
+				ScriptExt::Log(", ");
+
+			i++;
+		}
+
+		if (!pHouse->IsObserver())
+			ScriptExt::Log(" -> Main Enemy House: %d\n", pHouse->EnemyHouseIndex);
+		else
+			ScriptExt::Log("\n");
+	}
+}
```
This is probably no longer needed then either.

FS-21 replied:
Right, it was only for checking it was working as expected

Starkku reviewed (changes requested):
Some code style nitpicks, maybe try to reduce replicated code where possible.

Starkku commented:
Something I forgot to point out in the other scripting PR that got recently merged, the changelog page (What's New) contains a section called `For Map Editor (Final Alert 2)`. You should add entries for any new map trigger events & actions and AI script actions there.

FS-21 commented:
> Something I forgot to point out in the other scripting PR that got recently merged, the changelog page (What's New) contains a section called `For Map Editor (Final Alert 2)`. You should add entries for any new map trigger events & actions and AI script actions there.

Then I'll add the FA2 references from the previous map events PR here.

FS-21 committed:
Refactored code

ScriptExt::GetTheMostHatedHouse(...) was refactored due the multiple duplicated portions of code and other feedback changes-

FS-21 committed:
Small teak

FS-21 committed:
small tweak

FS-21 committed:
tweaks

Metadorius commented on src/Ext/Script/Body.AngerNodes.cpp:
```diff
@@ -0,0 +1,972 @@
+	if (!pTeam)
+		return;
+
+	for (auto& angerNode : pTeam->Owner->AngerNodes)
+	{
+		angerNode.AngerLevel = 0;
+	}
```
```suggestion
	for (auto& angerNode : pTeam->Owner->AngerNodes)
		angerNode.AngerLevel = 0;
```

FS-21 committed:
Update src/Ext/Script/Body.AngerNodes.cpp

Co-authored-by: Kerbiter <crabiter@vivaldi.net>

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/script-action-houses-anger

FS-21 committed:
FA2 documentation

FS-21 commented:
Added FA2 documentation of these script actions as requested.

Isn't nothing more left right?

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/script-action-houses-anger

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger

FS-21 committed:
Update YRpp

FS-21 committed:
Merge branch 'develop' into feature/script-action-houses-anger
